### PR TITLE
(GH-226) Use a dynamic pool_check loop period

### PR DESCRIPTION
**Pull request by glennsarti**
Previously the check_pool would always check the pool every 5 seconds, however
with a large number of pools, this can cause resource issues inside the
providers.  This commit:
- Introduces a dynamic check_pool period which increases during stability and
  decreases when the pool is being change in an important way
- Surfaces the settings as global config defaults but can also be set on a per
  pool basis
- Adds defaults to have default of 5 to 60 seconds and decay of 2.0
- Unit tests for the new behaviour

### CODE REVIEW

**Comment by glennsarti** [comment]
Interested in your opinion on this @mattkirby

**Comment by glennsarti** [comment]
As a side note, should the defaults be a more "sane" then e.g.  Default polling period is  (num of pools) / 4 so 20 pools = 5 sec polling period.

**Comment by mattkirby** [comment]
Yeah, I think it's definitely an improvement.

**Comment by glennsarti** [comment]
So my thinking was restricting the number of available connections doesn't stop how often a provider is polling, it just caps the maximum rate.  So it'll reach that hard limit fairly quickly when the number of connections is far lower than pools (as we've seen).  So I figured the one metric that has the greatest effect on provider load (at least in the vSphere instance) is how often vSphere is queried for the list of VMs in a Pool.

Another option would be to split operations that do and do not require an inventory scan.  This means we can limit the most "expensive" operation (inventory) to when we _actually_ need it e.g.
Inventory is needed:
- On the first pool scan for an initial inventory
- Just after a clone_vm (new vm will appear in the inventory)
- Just after a destroy_vm

In fact that ^^ may be a better way of doing it.  As my rolling timeout is still a little naive.

**Comment by mattkirby** [comment]
@glennsarti as it is I think this would be useful. In starting to work with this code with some of our vmpooler instances I find that I frequently hit the 60 second timeout. Is there anything I can do to help this one land? I'm all for further iteration, but I think this is useful as it stands now.

**Comment by glennsarti** [comment]
@mattkirby Sorry, been busy fixing other bits of infra.  I think I should change this to use more sane defaults.

**Comment by kevpl** [comment]
@glennsarti since you're already going to be changing the defaults, it seems better to make them into constants as well since you use them enough..

**Comment by glennsarti** [comment]
Okay, I've changed the defaults to 5 -> 60 and decay of 2.0.  Also used constants and updated docs.  Added some rubocop minor fixes for good measure.

**Comment by glennsarti** [comment]
Ready for merge.

**Comment by glennsarti** [comment]
... and tests have failed...

**Comment by glennsarti** [comment]
And tests are now good.  Ready for review agan

File: spec/unit/pool_manager_spec.rb [hunk 1]
```diff
@@ -2200,6 +2384,12 @@
         subject._check_pool(pool_object,provider)
       end
 
+      it 'should return the number of discoverd of VMs' do
```
Comment (shrug) [inline spec/unit/pool_manager_spec.rb]: nitpick: typo in "discovered"

Comment (glennsarti) [inline spec/unit/pool_manager_spec.rb]: Fixed

**Pull request merged**